### ISSUE TITLE
ExecStart=: do not consider slashes part of environment variable names

### ISSUE THREAD

**Issue by twojstaryzdomu** [systemd/systemd#27051]
### systemd version the issue has been seen with

252.6

### Used distribution

Debian

### Linux kernel version used

6.1.21+

### CPU architectures issue was seen on

x86_64

### Component

systemd

### Expected behaviour you didn't see

Systemd isn't stopping processing variables at the slash and appears to be considering slashes as part of the variable name. Slashes are frequently used in the shell to separate path elements (directories and files). When used in a shell variable with the `${var}` notation, a slash signifies a replacement e.g. `${var/old_term/new_term}`. Systemd ought to honour this handling of slashes and stop considering the slash, and any characters that follow it, as part of the variable name when resolving the variable.

Consider the service:
```
# cat /etc/systemd/system/test1.service
[Service]
Environment=FOO=/bin
ExecStart=/bin/ksh -c '$FOO/echo 123'
```
Had the slash not been considered for the variable name, it output should have been (identical to `/bin/ksh -c '/echo 123'):
```
ksh: /echo: not found
```
Alternatively, had the `$FOO` variable name been resolved, the output should have been `123`.

Mind any fix should not break the variable replacement notation in the shape `$${var/old_term/new_term}` or  `$${var/old_term/new_term}`, as evidenced in the sample below:

```
# cat /etc/systemd/system/test2.service
[Service]
Environment=FOO=123
ExecStart=/bin/ksh -c 'FOO=$${FOO/123/456}; echo $${FOO}'
```

should still render log output:
```
Jan 01:00:00 hostname systemd[1]: Started test2.service.
Jan 01:00:00 hostname ksh[1000]: 456
Jan 01:00:00 hostname systemd[1]: test2.service: Deactivated successfully.
```

Note this bug report is a call to fix the underlying slash handling problem, not call to merely warn users about it, as done in #27036.

### Unexpected behaviour you saw

Case in point, the systemd service:
```
# cat /etc/systemd/system/test1.service
[Service]
Environment=FOO=/bin
ExecStart=/bin/ksh -c '$FOO/echo 123'
```
produces the error `ksh: -c requires argument`, which indicates that the slash must have been considered part of the variable name.

```
Jan 01:00:00 hostname systemd[1]: Started test1.service.
Jan 01:00:00 hostname ksh[1000]: ksh: -c requires argument
```
which appears to have been the output of `ksh -c`.

### Steps to reproduce the problem

See earlier description (`test1.service` log output).

**Comment by poettering** [comment]
Warning makes sense, but just adding fuzzy ambiguous parsing doesn't make much sense to me. The syntax is clearly documented already, and people shouldn't use `foo${BAR}quux` syntax if they want the replacement to be done as part of a word.

We are not a bourne shell, and I don't think we should try to be one.

moreover the `$FOO` syntax generates a *series* of whitespace separated arguments, while `${FOO}` generates only one. This is a very clear rule. If we'd support `$FOO/quux` then this rule would be broken, because what would then even mean of FOO is set too `a b c`?

Anyway, totally not convinced we should change the parsing rules. Yes, we should warn about invalid env var names. But no we should not change the rules how we interpret them.

**Comment by twojstaryzdomu** [comment]
I'm honestly let down at such closure and not buying it at all.

>  people shouldn't use foo${BAR}quux syntax

But that works, doesn't it, surely you meant `foo$BARquux`?

Where did I state systemd is to be a shell? All I ask is to behave sensibly with variables and the different styles that users are used to. All that systemd's becoming now is warnings and a documentation nightmare (it's cheaper than fixing a bug), rather than fix the little bugs such as this to make it a better product. No matter how many spaces follow or precede, no amount of slashes is permitted in a variable.

My main concern is the lack of consistency. Why permit such syntax in the first place (`$FOO`) and stop half way through, accepting any characters, when you know well that some aren't proper candidates for a variable name? If you want to embrace ambiguous syntax like `$FOO`, fine, but then do not accept input ('/') that isn't likely to be processed at all as part of a variable name.

Users are understandably confused, which is why #27036 got raised. See bug #27039 about a similar lack of consistency in handling the little details between `Exec..` statements. Yes, you might close bugs arbitrarily without a fix, happily thinking a warning is enough or document every inconsistency, but those isn't making `systemd` any better. Users will still run into those pitfalls, the inconsitencies will remain.

Or perhaps the goal to not fixing those bugs is to give support teams something to do?

**Comment by yuwata** [comment]
(At least tentatively) reopen this. I will consider the report later.

**Comment by twojstaryzdomu** [comment]
Thank you for a second pair of eyes.

**Comment by YHNdnzj** [comment]
Why is this a "problem"? Environment variables themselves clearly accept slashes in names, and we should _block_ such usage.

**Comment by YHNdnzj** [comment]
Sure, I understand that this kind of var names are mostly broken, so I agree with the warning part. But since it's accepted by libc and kernel, let's don't bother to block it on our side.

**Comment by twojstaryzdomu** [comment]
If `systemd` accepts slashes, it is only by a glaring omission whilst implementing the parser, not by a conscious design decision, because then such use would be documented. Shells do not accept slashes in variable names historically to allow simple-form (`$FOO`) shell variables to be used in paths (e.g. `$DIR/$EXEC`). Why can't this be preserved for the sake of new users?

It is important to consider where new `systemd` users come from and ensure if one construct is allowed (`$FOO`), then that ambiguities play in (new) users' favour (e.g. slashes are not still not part of variable names).

I haven't heard a single argument what purpose allowing the slash iby `systemd` variable names serves. It's a bug, an omission of the designer didn't consider how many new users coming to `systemd` with their shell experiences are going to hit this problem.

In plain c, a slash signifies a division, not a valid variable name by the way.

**Comment by YHNdnzj** [comment]
> If systemd accepts slashes, it is only by a glaring omission whilst implementing the parser, not by a conscious design decision, because then such use would have be documented.
> ...
> In plain c, a slash signifies a division, not a valid variable name by the way.

Seriously?? Try to do `setenv("foo/bar", "test", 1)` in C or even `os.environ["foo/bar"] = "test"` for yourself.

**Comment by YHNdnzj** [comment]
> It is important to consider where new systemd users come from and ensure if one construct is allowed ($FOO), then that ambiguities play in (new) users' favour (e.g. slashes are not still not part of variable names).

I guess shell documents this because _it_ is the special one that do _not_ accept such syntax. So we just screw all other languages/parsers that accept this and use the shell syntax? _Why_?

**Comment by twojstaryzdomu** [comment]
In plain c, a slash is not a permitted character for a c variable.

Are these common examples or merely odd-ball ones? How are users typically going to access such variables without special escaping around them? For all intents purposes, slashes are most typical of paths, not obscure cases like those that you mention. Otherwise, they wouldn't be opening issues such as #27036. I understand wanting to cater to different audiences but given the woes of the environment variables in `systemd`, how many users will intentionally use the obscure syntax specifically to use a slash in a var name and how many will find their variable names around a slash fail due to their instinctive use of it from the shell?

If you allow slashes knowingly, it has to be stated in plain sight and consequences made clear that `systemd` considers EVERYTHING after a `$` to be part of a variable, unless `{}`s are involved. Oh and you do not 'screw' languages/parsers, those are inanimate, mind your manners.

**Comment by YHNdnzj** [comment]
> In plain c, a slash is not a permitted character for a c variable.

Yeah. But we're talking about environment variables here.

> How are users typically going to access such variables without special escaping around them? 

Through `getenv()`, as they are environment variables.

> I understand wanting to cater to different audiences but given the woes of the environment variables in systemd, how many users will intentionally use the obscure syntax specifically to use a slash in a var name and how many will find their variable names around a slash fail due to their instinctive use of it from the shell?

But I'm still against blocking it arbitrarily. Slashes and whitespaces are totally fine in env vars without any sort of escaping.

> If you allow slashes knowingly, it has to be stated in plain sight and consequences made clear that systemd considers EVERYTHING after a $ to be part of a variable, unless {}s are involved. Oh and you do not 'screw' languages/parsers, those are inanimate, mind your manners.

As long as we don't block this, I agree with all possible improvements including documentation/runtime warning. Sorry for being too aggressive.

**Comment by oficsu** [comment]
> We are not a bourne shell, and I don't think we should try to be one

@poettering, I agree that we're not a shell. But it's a user experience issue. Good UX must be at least one of the following: logical or consistent with other similar tooling. There is no logic to consider slashes (and other non-alphabetic symbols) a part of a variable. And at the same time it's not similar to other system tools 

Is it a really good idea to force users to read man pages and debug simple units rather than allow the tool to be simple and intuitive when a user does a simple task?

**Comment by twojstaryzdomu** [comment]
> But I'm still against blocking it arbitrarily. Slashes and whitespaces are totally fine in env vars without any sort of escaping.

You are correct, slashes are permitted though chances are very slim any user will rely on them in variable names, given the inherent ambiguity and the difficulty of escaping them. A question remains whether to allow that one obscure use or to cater to MOST users who will very likely hit this roadblock at some point. Yes warning is the right thing (e.g. Use `${FOO}` instead of `${FOO}` to dispel any ambiguity) though I'd go that extra step of excluding the space due to the sheer likelyhood anyone would venture to use it for the sake of the typical uses.

Is it possible to indicate to the user in whatever warning `systemd` would  produce what their shell variable rendered as because of the use of a slash? That would make it clearer why a shell issued a `requires an argument` warning or why it was not processed as a parameter, assuming shell-like syntax?
